### PR TITLE
feat(vite): error messages when .server files are referenced by client

### PR DESCRIPTION
**Pull request by pcattori**
https://github.com/remix-run/remix/assets/1477317/77eceb06-034a-4f6f-89a0-a2215686ada2

## TODO

- [x] changeset

### CODE REVIEW

File: integration/helpers/vite-template/tsconfig.json [hunk 1]
```diff
@@ -14,7 +14,9 @@
     "forceConsistentCasingInFileNames": true,
     "baseUrl": ".",
     "paths": {
-      "~/*": ["./app/*"]
+      "~/*": ["./app/*"],
+      "#dir/*": ["./app/.server/*"],
+      "#file": ["./app/utils.server"]
```
Comment (pcattori) [inline integration/helpers/vite-template/tsconfig.json:19]: this shouldn't be committed

File: packages/remix-dev/vite/plugin.ts [hunk 2]
```diff
@@ -927,22 +929,88 @@ export const remixVitePlugin: RemixVitePlugin = (options = {}) => {
...
+
+        throw Error(
+          [
+            colors.red(`Server-only module referenced by client`),
+            "",
+            `    '${id}' imported by '${importerShort}'`,
+            "",
+            `  * If all code in '${importerShort}' is server-only:`,
+            "",
+            `    Rename it to '${dotServerFile}'`,
```
Comment (markdalgleish) [inline packages/remix-dev/vite/plugin.ts:996]: I'm wondering if this should also mention `.server` directories?

Comment (pcattori) [reply packages/remix-dev/vite/plugin.ts:996]: This is meant to be a "quickfix" option to help unblock anyone migrating. We could maybe reword to make that clearer. That said, I can't think of a good way to mention `.server` directories that doesn't feel heavier. Any suggestions?

**Review by markdalgleish — APPROVED** [review]
This is so much nicer!
